google/api_core/retry_streaming.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Generator wrapper for retryable streaming RPCs.
+This function will be used when initilizing a retry with
+``Retry(is_stream=True)``.
+
+When ``is_stream=False``, the target is treated as a callable,
+and will retry when the callable returns an error. When ``is_stream=True``,
+the target will be treated as a callable that retruns an iterable. Instead
+of just wrapping the initial call in retry logic, the entire iterable is
+wrapped, with each yield passing through the retryable generator. If any yield
+in the stream raises a retryable exception, the entire stream will be
+retried.
+
+Important Note: when a stream is encounters a retryable error, it will
+silently construct a fresh iterator instance in the background
+and continue yielding (likely duplicate) values as if no error occurred.
+This is the most general way to retry a stream, but it often is not the
+desired behavior. Example: iter([1, 2, 1/0]) -> [1, 2, 1, 2, ...]
+
+There are two ways to build more advanced retry logic for streams:
+
+1. Wrap the target
+    Use a ``target`` that maintains state between retries, and creates a
+    different generator on each retry call. For example, you can wrap a
+    network call in a function that modifies the request based on what has
+    already been returned:
+
+    ```
+    def attempt_with_modified_request(target, request, seen_items=[]):
+        # remove seen items from request on each attempt
+        new_request = modify_request(request, seen_items)
+        new_generator = target(new_request)
+        for item in new_generator:
+            yield item
+            seen_items.append(item)
+
+    retry_wrapped = Retry(is_stream=True)(attempt_with_modified_request, target, request, [])
```

Comment:
Hmm, I'm not sure why this comment didn't make on my previous review:

`Retry` takes a unary target, so I would tweak the example as follows. WDYT?

```py
def attempt_with_modified_request(target, request):
    seen_items=[]
    def attempter(target):
        # remove seen items from request on each attempt
        new_request = modify_request(request, seen_items)
        new_generator = target(new_request)
        for item in new_generator:
            yield item
            seen_items.append(item)
    return attempter

retry_wrapped = Retry(is_stream=True)(attempt_with_modified_request(target, request))

```

Reply:
Good catch. That change looks fine to me, but I might have a slight preference to just breaking the end into two lines to avoid the extra nesting:

```python
    def attempt_with_modified_request(target, request, seen_items=[]):
        # remove seen items from request on each attempt
        new_request = modify_request(request, seen_items)
        new_generator = target(new_request)
        for item in new_generator:
            yield item
            seen_items.append(item)

    retry_wrapped_fn = Retry(is_stream=True)(attempt_with_modified_request)
    retryable_generator = retry_wrapped_fn(target, request)
```

Let me know what you think

Reply:
OK, I think that should work. Had to spend some time on the various levels of indirection, with the decorators and generators.

Reply:
Yeah I'm fine with either way, whatever seems more readable to you